bpf: convert IPV4_LOOPBACK to runtime variable, pass IPv4 addresses using union v4addr

jrife opened the pull request:
Since support for `NODE_CONFIG` was merged, revive the part of my older PR, #38022, that moved `IPV4_LOOPBACK` out of node_config.h. I wanted to start small for now to make sure I'm on the right track before migrating more config. This is part of https://github.com/cilium/cilium/issues/38370.

jrife committed:
bpf: declare IPV4_LOOPBACK as service_loopback_ipv4 runtime variable

Revive my older PR, #38022, since support for NODE_CONFIG was merged and
move IPV4_LOOPBACK out of node_config.h.

Signed-off-by: Jordan Rife <jrife@google.com>
Signed-off-by: Timo Beckers <timo@isovalent.com>

thorn3r reviewed (approved):
LGTM for agent changes

ti-mo commented on bpf/include/bpf/config/node.h:
```diff
@@ -15,4 +15,6 @@
 #include <node_config.h>
 
+NODE_CONFIG(__u32, ipv4_loopback, "IPv4 loopback address")
+#define IPV4_LOOPBACK CONFIG(ipv4_loopback)  /* Backwards compatibility */
```
IPV4_LOOPBACK only has a handful of references outside of tests, so I'd drop the macro and use `CONFIG(ipv4_loopback)` directly. The reason for some of the existing backwards compat macros is to avoid pinging all maintainers, but in this case we should be good.

Edit: looks like using __be32 for ipv4 addresses is very common in the bpf codebase, which may not be worth changing at the moment. In that case, I'd define `IPV4_LOOPBACK` as `bpf_ntohs(loopback_ipv4)` and make loopback_ipv4 a [4]byte.

Edit2: I'm confused by our use of `byteorder.NetIPv4ToHost32` in the agent to pass the value down to the datapath. This should mean the `__be32 saddr` fields actually contain LE values? Or maybe the Go value formatter changes the order of the bytes? This is one of those issues I'd like to sidestep with this work altogether..

jrife replied:
> IPV4_LOOPBACK only has a handful of references outside of tests, so I'd drop the macro and use CONFIG(ipv4_loopback) directly. The reason for some of the existing backwards compat macros is to avoid pinging all maintainers, but in this case we should be good.

Sure, I'll swap things out to use `CONFIG(ipv4_loopback)` explicitly (I wanted to make the diff small at first until someone reviewed -- I wasn't sure what the preference would be here).

> Edit: looks like using __be32 for ipv4 addresses is very common in the bpf codebase, which may not be worth changing at the moment. In that case, I'd define IPV4_LOOPBACK as bpf_ntohs(loopback_ipv4) and make loopback_ipv4 a [4]byte.

Will do.

jrife replied:
> Edit2: I'm confused by our use of byteorder.NetIPv4ToHost32 in the agent to pass the value down to the datapath. This should mean the __be32 saddr fields actually contain LE values? Or maybe the Go value formatter changes the order of the bytes? This is one of those issues I'd like to sidestep with this work altogether..

I believe that before, `WriteNodeConfig` used `NetIPv4ToHost32` when it wrote out the hex to `node_config.h` so that when the value was read back as a hex value it was correctly ordered in memory. I dropped its use altogether in the latest push, so a byte array should stay a byte array all the way through.

ti-mo commented on bpf/include/bpf/config/node.h:
```diff
@@ -15,4 +15,6 @@
 /* Legacy node config rendered at agent runtime. */
 #include <node_config.h>
 
+NODE_CONFIG(__u32, ipv4_loopback, "IPv4 loopback address")
```
A few thoughts:
- Paul made a list of [most all macros used in the datapath](https://github.com/cilium/cilium/issues/15379#issuecomment-897441189), I suggest we stick to those wherever possible, or at leat use them as a starting point.
- This can be turned into a `[4]byte` so we can do `node.LoopbackIPv4 = ([4]byte)(lnc.LoopbackIPv4)` below.
- Looking at other variables in the codebase, it's more common to use _ipv4 and _ipv6 as a suffix; don't feel obliged to retain the names of the macros you port over 1:1. For example, there's already `endpoint_ipv4/6` in lxc.h, which flows more naturally when read/pronounced.

jrife replied:
> Paul made a list of https://github.com/cilium/cilium/issues/15379#issuecomment-897441189, I suggest we stick to those wherever possible, or at leat use them as a starting point.

I'm a bit confused, are you referring to the description, `"IPv4 loopback address"`?

> Looking at other variables in the codebase, it's more common to use _ipv4 and _ipv6 as a suffix; don't feel obliged to retain the names of the macros you port over 1:1. For example, there's already endpoint_ipv4/6 in lxc.h, which flows more naturally when read/pronounced.

Does `loopback_ipv4` sound good to you instead?

ti-mo commented on bpf/tests/hairpin_sctp_flow.c:
```diff
@@ -28,6 +28,7 @@ mock_ctx_redirect_peer(const struct __sk_buff *ctx __maybe_unused, int ifindex _
 
 /* Set the LXC source address to be the address of pod one */
 ASSIGN_CONFIG(__u32, endpoint_ipv4, v4_pod_one)
+ASSIGN_CONFIG(__u32, ipv4_loopback, 0x1ffff50a)
```
Be skeptical of the dummies specified in `node_config.h`, they were mostly placeholders to keep code from getting optimized out. Is this strictly necessary here? Do the tests pass without it? Could we use a non-magic value here instead? (e.g. v4_node_one)

jrife replied:
Tests don't pass without this. Before, they were getting the value from the defaults defined in `node_config.h` I believe. I can define `v4_loopback` or some such macro instead here.

ti-mo commented on pkg/datapath/loader/config.go:
```diff
@@ -12,6 +13,10 @@ import (
 func nodeConfig(lnc *datapath.LocalNodeConfiguration) config.Node {
 	node := *config.NewNode()
 
+	if option.Config.EnableIPv4 {
```
I'd hoped the `CiliumInternalIPv6` would've inspired folks to take the same approach. :wink: All addresses can (should!) be turned into arrays now to avoid issues with endianness, unnecessary conversions and ambiguity.

jrife replied:
Changed in latest push.

ti-mo reviewed (changes requested):
Thanks for pushing this forward! Left a few comments. This is great to get our feet wet so we can figure out some conventions.

ti-mo committed:
*: rename LoopbackIPv4 to ServiceLoopbackIPv4

Naming things is hard. LoopbackIPv4 is one of those variables I've always
been confused by. Its use and meaning was documented in some places, but
the name was always somewhat of a misnomer.

It's not an address bound to a loopback interface. It's not a valid
destination address for a packet. It's an internal implementation detail
that sometimes needs to be overridden when it overlaps with an in-use
address range.

This commit replaces all occurrences of LoopbackIPv4 with ServiceLoopbackIPv4
and extends docstrings wherever the existing doc was too terse.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: remove LXC_IPV4 macro, convert to union v4addr

From the agent's side, we want to avoid endianness conversions since they make
the end result hard to reason about. Convert endpoint_ipv4 to union v4addr to
allow the agent to provide netip.Addr bytes.

Substitute the LXC_IPV4 macro with CONFIG(endpoint_ipv4).

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: remove IPV4_MASQUERADE macro, convert to union v4addr

Like the previous commit, this converts nat_ipv4_masquerade to union v4addr
and removes the backwards-compat macro.

Signed-off-by: Timo Beckers <timo@isovalent.com>
